### PR TITLE
Correct MathText color, size and vertical alignment

### PR DESCRIPTION
**Pull request by IuryPiva**
# Note
As planned on https://github.com/bokeh/bokeh/discussions/11720 this is ready to be merge into `mathtext-main-branch` without the baselines, the baselines will be added on `mathtext-main-branch` once we have a deliverable ready. 

The baselines have been download and checked to ensure this works correctly.

-----
Current display of mathtext have different colors, sizes and vertical alignment compared to other Bokeh text. Since we can't do much to change the actual [font MathJax outputs](http://docs.mathjax.org/en/latest/output/fonts.html), we can automatically add macros to match the color from the model it's been used on, and adjust the size and vertical alignment making roughly the same calculations a CSS engine would do if the SVG was displayed on a HTML file.

Currently, if you use 
```python
p.yaxis.axis_label = r"\[\sin(x)\]"
p.yaxis.text_color = "white"
```
MathJax will not know about the color property and the text generated by MathJax will appear in MathJax' default color, unless you explicitly include a MathJax macro, like `r"$${\color{white} \sin(x)}$$"`. This PR will enable styling for MathText, i.e. translate properties into MathJax macros. Only `text_color`, `text_font_size` will be available and more work should be done to respect all Bokeh's standard text properties (https://docs.bokeh.org/en/latest/docs/user_guide/styling.html#userguide-styling-text-properties).

A manually added macro overrides an object property, i.e:
```python
p.yaxis.axis_label = r"$${\color{white} \sin(x)}$$"
p.yaxis.text_color = "green"
```

Will result in a white text, not green.

As of vertical positioning, when we display MathJax on a normal web page it looks like this:
image
Now, if I give it a border we can see that there is some padding and some vertical alignment that display each element on roughly the alphabetical baseline:
image
<details>
  <summary>With console open:</summary>
  open console
</details>

We can calculate the same position on Canvas with the correct font metrics, [counting for line height](https://stackoverflow.com/questions/23247467/how-css-line-height-is-measured) and applying the vertical alignment numbers that MathJax includes in the `style` property of the outputted SVG.

## Results:
<details>
  <summary>Before:</summary>
  
  Screen Shot 2021-09-18 at 10 22 57
</details>
<details>
  <summary>After:</summary>
  
  Screen Shot 2021-09-18 at 11 52 15
</details>

<details><summary>On vertical align</summary>
Note: The 10^8 and other 10s are a custom graphics box called BaseExpo, it is not text nor MathText. We probably should change them too.

log axis side by side

log axis side by side anchor top

</details>
<details><summary>On title annotations</summary>
Note: The prefixed text (Above, Left, ...) is inside a \text macro and is overriding the bold macro. Should be fixed on #11734 

log axis side by side

</details>

- [x] relates to: #11355
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by bryevdv** [comment]
> Only text_color, text_font_size will be available and more work should be done to respect all Bokeh's standard text properties

I am not sure it will be possible to support *all* the text properties. That's fine, I think most LaTeX users will be well-aware that LaTeX likes to control much of its own rendering. We just need to be clear about what exactly is and is not expected to work together cc @tcmetzger

**Comment by bryevdv** [comment]
Regarding the vertical spacing, I'll take your word. I tried, but I basically can't see the difference before/after. If there was a side-by-side image comparison, or actual image diff, it would probably be easier for me.

**Comment by IuryPiva** [comment]
> Regarding the vertical spacing, I'll take your word. I tried, but I basically can't see the difference before/after. If there was a side-by-side image comparison, or actual image diff, it would probably be easier for me.

Oh sorry, I forgot to upload images of that. The changes are noticeable on tick labels. Just added a diff screenshot under Results > "On vertical align" on the description.

**Comment by mattpap** [comment]
This should have been targeted for 3.0 and then we would decide if it makes sense to backport it, not the other way round.

**Comment by bryevdv** [comment]
> This should have been targeted for 3.0 and then we would decide if it makes sense to backport it, not the other way round.

@mattpap this is a reasonable topic for discussion but AFAIK there is not a decision on it. As far as I am concerned all the CZI work should definitely be part of 2.4.x releases, which make fwd vs back port largely a matter of preference.

**Comment by mattpap** [comment]
> As far as I am concerned all the CZI work should definitely be part of 2.4.x releases (...)

This is not what patch releases are for, i.e. no major new development. As far I'm concerned all new development, in particular new development related to math text support, will be a part of 3.0 release.

**Comment by bryevdv** [comment]
> As far I'm concerned all new development, in particular new development related to math text support, will be a part of 3.0 release

I guess we will have to have a formal decision because I absolutely do not support waiting that long to land more math text work.

**Comment by IuryPiva** [comment]
I think this one is a "visual fix", to look like it's is supposed to, not really a new feature. It would make sense for this one to make it into 2.4.1

**Comment by IuryPiva** [comment]
This PR should wait on a new PR where I will organize our current MathText integration tests. Each test on Axis integration test file generate 8 baselines, one for each side and one for LogAxis other for LinearAxis, and even for things that should not matter where they are positioned.

**Comment by tcmetzger** [comment]
Updated docs based on the changes in this PR are in #11660.

**Comment by IuryPiva** [comment]
@mattpap the visual tests are breaking because it briefly tries to fit a unformatted text before the image has loaded which will break the bbox max size 
image

My suggestion is to remove this display of text, it has no use, if the user sees a non processed tex string or a blank image he will know something went wrong.

**Comment by IuryPiva** [comment]
Just a note: These LaTeX PRs are a possible good candidate for https://github.blog/changelog/2021-10-27-pull-request-merge-queue-limited-beta/ put that still is in closed beta.

**Comment by bryevdv** [comment]
I have requested to join the waitlist for early access

File: bokehjs/src/lib/models/text/math_text.ts [hunk 1]
```diff
@@ -187,7 +187,8 @@ export abstract class MathTextView extends BaseTextView implements GraphicsBox {
   size(): Size {
     const {width, height} = this._size()
     const {angle} = this
-    if (!angle)
+
+    if (angle == null || angle == 0 || typeof angle == "undefined")
```
Comment (mattpap) [inline bokehjs/src/lib/models/text/math_text.ts:191]: Do you need to handle `undefined` here? `angle == null` already handles that. Besides, if you need to distinguish `null` from `undefined`, then you should really use symbols instead.

**Comment by mattpap** [comment]
Some of the baseline line images are incorrect, e.g.:
image
Each baseline has to be reviewed before commiting (even if a particular PR is not responsible for e.g. undeterministic behavior)..

**Comment by IuryPiva** [comment]
> Some of the baseline line images are incorrect, e.g.: image Each baseline has to be reviewed before commiting (even if a particular PR is not responsible for e.g. undeterministic behavior).

I'm sorry this slip through, I was supposed to rename this text to 'don't show anything if provider not available`.

I removed the text fallback as it was messing with baselines size, because it would need to have space for the fallback text that was never displayed. And if you are a user and see a text of your formula or an empty image you will have to search for what is going wrong anyway.

I will rename the test if you are okay with this new behavior

**Comment by mattpap** [comment]
The idea has been that if mathjax is not available, we render plain text. If there's an issue with testing this, then we need to address that, and not change the design.

**Comment by bryevdv** [comment]
> The idea has been that if mathjax is not available, we render plain text.

The latest baselines display a truncated ascii version in case mathtext does not render. What is the desired behavior? I tend to agree that no users are asking for long ascii math strings, so truncation is reasonable, mostly as an implicit error message. But I might be convinced the current truncation is too short.

**Comment by IuryPiva** [comment]
> The latest baselines display a truncated ascii version in case mathtext does not render. What is the desired behavior? I tend to agree that no users are asking for long ascii math strings, so truncation is reasonable, mostly as an implicit error message. But I might be convinced the current truncation is too short.

The fallback mechanism is correctly addressed and corrected on the next PR https://github.com/bokeh/bokeh/pull/11846 . Since we ditched CDN support for MathJax I can now synchronously calculate the size of the output image, which is why I had to truncate the text here.
